### PR TITLE
Stubbed in some ideas

### PR DESCRIPTION
**Pull request by jlaura**

### CODE REVIEW

File: knoten/csm.py [hunk 1]
```diff
@@ -18,192 +21,128 @@ def default(self, obj):
...
+    @intersect.register(GeoDataset)
+    def _(self, dem, image_pt, tolerance=1, max_its=20):
+        if not isinstance(image_pt, csmapi.ImageCoord):
+            # Support a call where image_pt is in the form (x,y)
+            image_pt = csmapi.ImageCoord(*image_pt)
+
+        intersection = self.camera.imageToGround(image_pt, 0.0)
+        iterations = 0
+        while iterations < max_its:
+            height =
```
Comment (acpaquette) [inline knoten/csm.py:96]: @jlaura Uh, what was this supposed to be? It looks like it might be in an uncommited/pushed code change

Comment (jessemapel) [reply knoten/csm.py:96]: This should be the height from the dem at the current intersection lat, lon

File: knoten/csm.py [hunk 2]
```diff
@@ -18,192 +21,128 @@ def default(self, obj):
...
+        while iterations < max_its:
+            height = 
+            next_intersection = camera.imageToGround(image_pt, height)
+            dist = max(abs(intersection.x - next_intersection.x),
+                    abs(intersection.y - next_intersection.y),
+                    abs(intersextion.z - next_intersection.z))
+
+            intersection = next_intersection
+            if dist < tolerance:
+                return intersection
```
Comment (jessemapel) [inline knoten/csm.py:104]: Returning None when we do not converge sounds hard to trace back down. We should either return what we have an a warning or raise and exception.

File: knoten/warp.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import jinja2
+
+def generate_vrt(raster_size, gcps, fpath,
```
Comment (jessemapel) [inline knoten/warp.py:3]: I'd rather call this to_vrt or something else to indicate it's writing out a file.

File: knoten/csm.py [hunk 4]
```diff
@@ -18,192 +21,128 @@ def default(self, obj):
...
+        return ellipsoid.getSemiMinorRadius()
+
+    @methodispatch
+    def intersect(self, height, image_pt):
+        if not isinstance(image_pt, csmapi.ImageCoord):
+            image_pt = csmapi.ImageCoord(*image_pt)
+        return self.model.imageToGround(image_pt, height)
+
+    @intersect.register(GeoDataset)
+    def _(self, dem, image_pt, tolerance=1, max_its=20):
```
Comment (jessemapel) [inline knoten/csm.py:88]: The default tolerance should be a function of the ground pixel size. We there should be a way to compute this from the output of computeGroundPartials, it is just slipping my mind at the moment.

File: knoten/csm.py [hunk 5]
```diff
@@ -18,192 +21,128 @@ def default(self, obj):
...
+        if not isinstance(image_pt, csmapi.ImageCoord):
+            # Support a call where image_pt is in the form (x,y)
+            image_pt = csmapi.ImageCoord(*image_pt)
```
Comment (jessemapel) [inline knoten/csm.py:91]: A ground to image wrapper that does a similar conversion for x, y, z or lat, lon would be very helpful.

**Comment by jlaura** [comment]
With @acpaquette's PR in, I am going to close this until we all have a larger discussion about the architecture of knoten.

**Pull request closed without merging**